### ISSUE TITLE
Missing operations for `half_t` on HIP

### ISSUE THREAD

**Issue by aprokop** [kokkos/kokkos#6404]
**Describe the bug**

The following code
```c++
Kokkos::parallel_for(1, KOKKOS_LAMBDA(int) {
  Kokkos::Experimental::half_t x;
    0.f < x; // (1)
    auto y = Kokkos::sqrt(x); // (2)
});
```
results in an errors on HIP:
```
(1)
invalid operands to binary expression ('float' and 'Kokkos::Experiment
al::half_t' (aka 'floating_point_wrapper<__half>'))
    0.f < x;
    ~~~ ^ ~
(2)
no matching function for call to 'sqrt'
    auto y = Kokkos::sqrt(x);
             ^~~~~~~~~~~~
kokkos/include/Kokkos_MathematicalFunctions.hpp:376:33: note:
candidate function not viable: no known conversion from 'Kokkos::Experimental::half_t' (aka 'floating_point_wrapper<__half>') to 'float' for 1st argument
KOKKOS_IMPL_MATH_UNARY_FUNCTION(sqrt)
                                ^
```
I think (1) is a bug, and (2) may be a missing feature.

**Please also include the following items to support reproducing the bug**
- ROCm 5.3
- Kokkos `develop` (393abd3ef)
- AMD MI250X
- [KokkosCore_config.h.log](https://github.com/kokkos/kokkos/files/12504149/KokkosCore_config.h.log)
- [configure.sh.log](https://github.com/kokkos/kokkos/files/12504146/configure.sh.log)

**Comment by dalg24** [comment]
Math functions are coming in #6124 

The comparison failure is a bug though.  For the record `x < 0.f` compiles but `0.f < x` yields an error.

**Comment by e10harvey** [comment]
> For the record `x < 0.f` compiles but `0.f < x` yields an error.

The compiler is using our implicit conversion constructor to downcast `0.f` to half in the expression `x < 0.f`. But, this expression should upcast x to float. I'll add overloads to ensure this to #6407.

**Comment by masterleinad** [comment]
Fixed in #6407 (at least partly for this issue but not comparison with `int` etc.) and #6124.
